00X-ABI-Stable-Module-API.md

```diff
@@ -0,0 +1,377 @@
+
+### Types
+```C
+typedef struct napi_env__ *napi_env;
+typedef struct napi_value__ *napi_value;
+typedef struct napi_persistent__ *napi_persistent;
+typedef struct napi_handle_scope__ *napi_handle_scope;
+typedef struct napi_escapable_handle_scope__ *napi_escapable_handle_scope;
+typedef struct napi_propertyname__ *napi_propertyname;
+typedef const struct napi_func_cb_info__ *napi_func_cb_info;
+typedef void (*napi_callback)(napi_env, napi_func_cb_info);
+typedef void napi_destruct(void* v);
+
+
+struct napi_method_descriptor {
+  napi_callback callback;
+  char* utf8name;
+};
+
+extern "C" {
+enum napi_valuetype {
+    // ES6 types (corresponds to typeof)
+    napi_undefined,
+    napi_null,
+    napi_boolean,
+    napi_number,
+    napi_string,
+    napi_symbol,
+    napi_object,
+    napi_function,
+};
+```
+### Functions
+
+```C
+// Getters for defined singletons
+NODE_EXTERN napi_value napi_get_undefined(napi_env e);
+NODE_EXTERN napi_value napi_get_null(napi_env e);
+NODE_EXTERN napi_value napi_get_false(napi_env e);
+NODE_EXTERN napi_value napi_get_true(napi_env e);
+NODE_EXTERN napi_value napi_get_global_scope(napi_env e);
+
+
+// Methods to create Primitive types/Objects
+NODE_EXTERN napi_value napi_create_object(napi_env e);
+NODE_EXTERN napi_value napi_create_array(napi_env e);
+NODE_EXTERN napi_value napi_create_array_with_length(napi_env e, int length);
+NODE_EXTERN napi_value napi_create_number(napi_env e, double val);
+NODE_EXTERN napi_value napi_create_string(napi_env e, const char*);
+NODE_EXTERN napi_value napi_create_string_with_length(napi_env e, const char*,
+                                                               size_t length);
+NODE_EXTERN napi_value napi_create_boolean(napi_env e, bool b);
+NODE_EXTERN napi_value napi_create_symbol(napi_env e, const char* s = NULL);
+NODE_EXTERN napi_value napi_create_function(napi_env e, napi_callback cbinfo);
+NODE_EXTERN napi_value napi_create_error(napi_env e, napi_value msg);
+NODE_EXTERN napi_value napi_create_type_error(napi_env e, napi_value msg);
+
+
+// Methods to get the the native napi_value from Primitive type
+NODE_EXTERN napi_valuetype napi_get_type_of_value(napi_env e, napi_value v);
+NODE_EXTERN double napi_get_number_from_value(napi_env e, napi_value v);
+NODE_EXTERN int napi_get_string_from_value(napi_env e, napi_value v,
+                                           char* buf, const int buf_size);
+NODE_EXTERN int32_t napi_get_value_int32(napi_env e, napi_value v);
+NODE_EXTERN uint32_t napi_get_value_uint32(napi_env e, napi_value v);
+NODE_EXTERN int64_t napi_get_value_int64(napi_env e, napi_value v);
+NODE_EXTERN bool napi_get_value_bool(napi_env e, napi_value v);
+
+NODE_EXTERN int napi_get_string_length(napi_env e, napi_value v);
+NODE_EXTERN int napi_get_string_utf8_length(napi_env e, napi_value v);
+NODE_EXTERN int napi_get_string_utf8(napi_env e, napi_value v,
+                                     char* buf, int bufsize);
```

Comment:
opened https://github.com/nodejs/abi-stable-node/issues/25 to update current implementation.